static/js/compose_actions.js

```diff
@@ -98,6 +113,29 @@ function show_box(msg_type, opts) {
     set_focus(msg_type, opts);
 }
 
+let togglable_message_type;
+
+function toggle_message_type(opts) {
+    togglable_message_type = togglable_message_type === "private" ? "stream" : "private";
+    compose_state.set_message_type(togglable_message_type);
+    opts.message_type = togglable_message_type;
+    opts.trigger = "toggle message";
+    opts.stream = compose_state.stream_name();
+    opts.topic = compose_state.topic();
+    opts.private_message_recipient = compose_state.private_message_recipient();
+    reset_message_type_switcher(opts);
+    show_box(togglable_message_type, opts);
+    update_placeholder_text();
+}
+
+function reset_message_type_switcher(opts) {
+    // remove any previously set click handler and reset with new opts
+    $("#message_type_icon").off("click");
+    $("#message_type_icon").on("click", () => {
+        toggle_message_type(opts);
+    });
```

Comment:
What's the purpose of having `opts` saved for `toggle_message_type`?  Is the idea that `show_box` needs some options passed to it?  

I would kinda expect that whatever options were passed when originally opening the compose box would be irrelevant once you're switching the message type, but I've forgotten what those options are. It's possible we just want the switcher to pass a fixed set of options.

Reply:
Yup, it's passed through because `show_box` needs it for various things -- it determines if PM or Stream compose are shown, it also sets the focus which needs the rest of the `opts`.

I wish the options were typed, but I'm pretty sure it's always

```
{
    message_type,
    trigger
    stream
    topic
    private_message_recipient
}
```

Though I actually totally reset them `toggle_message_type`, so I don't need to pass the opts in there :)

It might be possible to just get rid of `opts`, I'll play around with that.